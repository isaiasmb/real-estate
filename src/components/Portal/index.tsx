import { useState, useEffect } from 'react'
import { createPortal } from 'react-dom'

type PortalProps = {
  children: React.ReactNode
}

const Portal = ({ children }: PortalProps) => {
  const [mount, setMount] = useState<HTMLElement | null>(null)
  useEffect(() => {
    const portalRoot = document.getElementById('portal-root')

    if (!portalRoot) {
      const newPortal = document.createElement('div')
      newPortal.setAttribute('id', 'portal-root')
      document.body.appendChild(newPortal)
    }

    setMount(document.getElementById('portal-root'))
  }, [])

  const el = document.createElement('div')

  useEffect(() => {
    if (!mount) {
      return
    }

    mount.appendChild(el)
    return () => {
      mount.removeChild(el)
    }
  }, [el, mount])

  return createPortal(children, el)
}

export default Portal
