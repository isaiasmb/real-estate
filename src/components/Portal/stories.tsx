import { Story, Meta } from '@storybook/react/types-6-0'
import Portal from '.'

export default {
  title: 'Portal',
  component: Portal
} as Meta

export const Default: Story = () => {
  return (
    <div>
      <Portal>
        <h1>Portal text 1</h1>
      </Portal>
      <Portal>
        <h1>Portal text 2</h1>
      </Portal>
    </div>
  )
}
