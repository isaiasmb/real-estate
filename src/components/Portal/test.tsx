import { render, screen } from '@testing-library/react'

import Portal from '.'

describe('<Portal />', () => {
  let consoleSpy: jest.SpyInstance

  beforeEach(() => {
    if (consoleSpy) {
      consoleSpy.mockRestore()
    }
  })

  it('should render the portal with children', () => {
    const { container } = render(
      <Portal>
        <h1>Portal Text</h1>
      </Portal>
    )
    expect(
      screen.getByRole('heading', { name: /Portal Text/i })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the portal only once', async () => {
    consoleSpy = jest.spyOn(console, 'error').mockImplementation()

    const { container } = render(
      <div>
        <Portal>
          <h1>Portal Text 1</h1>
        </Portal>
        <Portal>
          <h1>Portal Text 2</h1>
        </Portal>
      </div>,
      { container: document.body }
    )

    expect(
      screen.getByRole('heading', { name: /Portal Text 1/i })
    ).toBeInTheDocument()
    expect(
      screen.getByRole('heading', { name: /Portal Text 1/i })
    ).toBeInTheDocument()
    expect(container.querySelectorAll('#portal-root')).toHaveLength(1)
  })
})
