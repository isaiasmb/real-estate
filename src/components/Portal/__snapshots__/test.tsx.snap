// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portal /> should render the portal with children 1`] = `null`
